plugins/pulp_rpm/plugins/db/models.py

```diff
@@ -1005,6 +1011,33 @@ def verify_size(self, location):
             with open(location) as fp:
                 verification.verify_size(fp, self.size)
 
+    def set_repodata(self, metadata_type, xml_snippet):
+        """
+        Compress metadata and put it into `repodata` attribute which will be saved to the db later.
+
+        :param metadata_type: key for the `repodata` dictionary which indicates type of metadata
+        :type  metadata_type: str
+        :param xml_snippet: utf-8 string which will be compressed and put into `repodata` dictionary
+        :type  xml_snippet: str
+        """
+        if metadata_type in constants.MANDATORY_METADATA_TYPES:
+            if isinstance(xml_snippet, unicode):
+                xml_snippet = xml_snippet.encode('utf-8')
+            self.repodata[metadata_type] = bson.binary.Binary(gzip.zlib.compress(xml_snippet))
+
+    def get_repodata(self, metadata_type):
```

Comment:
I like your initial idea to do everything one way. Maybe we can test more on several customer-like setups to be more confident?
I also do not have strong objections to compress metadata for the newly added units only. But we have to keep in mind and not forget that we have two types of metadata:
- for any later migrations touching RPM metadata 
- for downstream cherry-picked migrations probably
- something else?

From all the proposed converting options I still think that migration is the safest one. Stand-alone tool or on-demand will do conversion on a running system I guess, so I would not rule out possible race conditions.